### ISSUE TITLE
Add ability to disable addons remotely for all users

### ISSUE THREAD

**Issue by WorldLanguages** [ScratchAddons/ScratchAddons#5248]
### Why this would be helpful/solve a problem

There are cases where we'd want to disable addons for all users without shipping an update.

Examples:
- Security issue (such as XSS) found in an a website addon
- Scratch editor update breaks an editor addons, and the update causes the editor to crash in situations which aren't unlikely
- Disable an April fools addon until it's April Fools
- Not showing addons in the settings page until an exact date

### What this change does

The background page would fetch `https://data.scratchaddons.com/disabledaddons.json` every 30 minutes.
Those addons that are marked as disabled should be logically disabled, but ideally the addon shouldn't be turned off in the actual user settings. We don't want an incorrect disabledaddons file to disable all addons for all users - they should be logically disabled, but the user's `chrome.storage.sync` should stay unchanged.

This file would include:
- The addon ID
- Affected version numbers

### Possible alternatives

I don't think there's a way that doesn't involve pinging a server every a certain amount of time.

**Comment by BroJac5246** [comment]
This should probably be added for security purposes- just promise that this won't be used to turn off April Fools' Day Addons after April Fools' Day 😉 (if you do want to do that, have the extension automatically disable them the day after April Fools' Day)

**Comment by mxmou** [comment]
> * Disable an April fools addon until it's April Fools
> * Not showing addons in the settings page until an exact date

That can be done without a server.

**Comment by WorldLanguages** [comment]
> That can be done without a server.

Only if we set the date beforehand. It can sometimes be tricky because we don't know how long the review process in the extension stores will take.

@mxmou Any feedback about the other examples? Those in particular are just random ideas, but these 2 are the main ones:



> * Security issue (such as XSS) found in an a website addon
> * Scratch editor update breaks an editor addon, and the update causes the editor to crash in situations which aren't unlikely

**Comment by DNin01** [comment]
I think for vulnerabilities, SA should **warn the user** that an addon with a vulnerability is enabled, and prompt them to disable it. I'm not sure what to do if an addon breaks Scratch.

**Comment by penguinmoose** [comment]
@DNin01 (just saw you commenting!) But that might take time - the user may ignore the warning. Instead, I think there should be a notice that lets the user know that one of their addons has been disabled because of ___.

**Comment by penguinmoose** [comment]
@WorldLanguages Instead of pinging a server constantly, I think it would be better if it only pings the server whenever a Scratch page that **involves the addon** is opened. There would be a timeout so Scratch Addons does not ping the server too often.

**Comment by penguinmoose** [comment]
Also, I don't think April Fools addons should be disabled until it's that day. Some people may want that addon on all the time.

**Comment by DNin01** [comment]
> I think there should be a notice that lets the user know that one of their addons has been disabled because of ___.

Oh yeah, _automatically_ disable the addon and let the user know that it was disabled.

**Comment by mxmou** [comment]
> Scratch editor update breaks an editor addon, and the update causes the editor to crash in situations which aren't unlikely

Would it be possible to implement an easier way to test SA in the development version of scratch-www?

> Security issue (such as XSS) found in an a website addon

I think we should find a way to solve this without pinging a server periodically, at least not in the background.

**Comment by lisa-wolfgang** [comment]
Related: https://github.com/ScratchAddons/ScratchAddons/issues/2620

**Comment by WorldLanguages** [comment]
> I think for vulnerabilities, SA should **warn the user** that an addon with a vulnerability is enabled, and prompt them to disable it. I'm not sure what we should do for the case that an addon breaks Scratch.

I don't think that makes sense. If a user really wants to use a vulnerable addon, they can remove the code that disables addons remotely (don't forget SA is open source).
If they don't know how to do that, then I'd say they probably don't understand the vulnerability. A user wouldn't know what XSS is if they don't have some basic coding experience.

**Comment by WorldLanguages** [comment]
> I think it would be better if it only pings the server whenever a Scratch page that **involves the addon** is opened.

The extension can't know in advance what the vulnerable/broken addon is, so I'm not sure what you mean by "involves **the** addon".

**Comment by WorldLanguages** [comment]
> Would it be possible to implement an easier way to test SA in the development version of scratch-www?

Yes! Could you create an issue and tell us how you've been doing it until now, and how you'd wish it was? (I think you mentioned how you've been doing it somewhere at some point, but I don't know how to find it).

**Comment by WorldLanguages** [comment]
> I think we should find a way to solve this without pinging a server periodically, at least not in the background.

I can't think of any other way.
Could you clarify what you feel is the concern here? Network usage due to periodic pinging? Privacy issues? Feels sketchy?

**Comment by Secret-chest** [comment]
Privacy concerns.

**Comment by Secret-chest** [comment]
> Not showing addons in the settings page until an exact date

Why? If you think it's a good idea, it's forking time.

**Comment by WorldLanguages** [comment]
The SA privacy policy has always considered pinging ScratchAddons.com as something the extension could do:

https://scratchaddons.com/docs/policies/privacy/extension/#third-parties

> By default, Scratch Addons only reaches 2 services automatically:
> 1. The Scratch website (for example, to get your unread message count).
> 2. This website (for example, https://scratchaddons.com/welcome opening as a new tab after you’ve installed the extension).

Until now, scratchaddons.com was only reached automatically through `chrome.tabs.create()`, but using `fetch()` would also be considered "reaching a service automatically".

**Comment by WorldLanguages** [comment]
Regarding privacy, I suggest:
- Creating a dedicated subdomain for this, http://data.scratchaddons.com
- Not proxying this subdomain through Cloudflare, and point the DNS directly to a public GitHub repository through GitHub Pages. GitHub would be the only middleman here. As the owner of the domain, I wouldn't be able to see the requests coming in because this is a simple CNAME record (basically a DNS redirect).

**Comment by mxmou** [comment]
> Regarding privacy, I suggest:
> 
> * Creating a dedicated subdomain for this, https://data.scratchaddons.com
> * Not proxying this subdomain through Cloudflare, and point the DNS directly to a public GitHub repository through GitHub Pages. GitHub would be the only middleman here. As the owner of the domain, I wouldn't be able to see the requests coming in because this is a simple CNAME record (basically a DNS redirect).

I think it's also very important to not ping too frequently and only when the extension is being actively used (for example when visiting the Scratch website).

**Comment by WorldLanguages** [comment]
> I think it's also very important to not ping too frequently

Okay, can you define "frequently"? Also, why?

> only when the extension is being actively used (for example when visiting the Scratch website)

What you said we should do wouldn't have any downsides, because it would only make sense to remotely disable userscripts. No userscripts will run if the user doesn't visit Scratch. So, that is OK.

**Comment by DNin01** [comment]
> Could you clarify what you feel is the concern here? Network usage due to periodic pinging? Privacy issues? Feels sketchy?

For me, I'm mostly thinking about resource usage here. It can add up for some users.

**Comment by lisa-wolfgang** [comment]
I wouldn't expect that a single request would take up any significant amount of data. I would be surprised if it exceeded 1MB per day.

**Comment by WorldLanguages** [comment]
As I said above, we could avoid fetching the list of disabled addons unless the user has opened Scratch.
The size of the file would be insignificant compared to network/resource usage from Scratch itself.

**Comment by lisa-wolfgang** [comment]
So what will that look like from a technical standpoint? Would we defer Scratch from loading until we get a response from scratchaddons.com? If scratchaddons.com is down, that would make opening a new Scratch tab several seconds slower every half hour. Another option is to defer loading addon userscripts until this request is received, but that has the same drawbacks as deferring Scratch. Yet another option is to dynamically disable any affected addons once the response is received. However, if we let Scratch load first and then get the response saying to disable a vulnerable addon, we would likely reload the page, but the affected userscript might have injected itself at that point if the user's connection to scratchaddons.com was slow for whatever reason.

**Comment by WorldLanguages** [comment]
@lisa-wolfgang Those are nice questions that I wanted to ask myself (and possibly here) at some point.

> Would we defer Scratch from loading until we get a response from scratchaddons.com? 

I think we shouldn't, at least on the first implementation. Also, if scratchaddons.com is down, the extension would work normally and continue to use the old list of disabled addons (until the extension gets updated to a new version, the moment where the cache should be cleared)

> we would likely reload the page

It's probably a bad idea to reload the page forcefully.
Disabling a vulnerable/broken addon after 1 refresh is better than having to wait multiple hours for extension store review anyway.
We shouldn't be too strict about our first implementation of remote disabling.

**Comment by yzffkindacrzy** [comment]
This honestly isn't that bad of an idea, althought is at the same time, heres what I am trying to say:

If you were able to remotely disable addons for all users, this could cause trouble and will require a server.
Or if you dont want to use a server you could also use a github gist for the extension to fetch a json file with enabled and disabled extensions, but you probably shouldn't do this since users could be editing and extensions all get disabled and they may have to renable all there extensions again, so If I were you I probably wouldn't.

**Comment by DNin01** [comment]
> This honestly isn't that bad of an idea, althought is at the same time, heres what I am trying to say:
> 
> If you were able to remotely disable addons for all users, this could cause trouble and will require a server. Or if you dont want to use a server you could also use a github gist for the extension to fetch a json file with enabled and disabled extensions, but you probably shouldn't do this since users could be editing and extensions all get disabled and they may have to renable all there extensions again, so If I were you I probably wouldn't.

It wouldn't have to disable _all_ of the addons, just the ones with vulnerable code, for example.

**Comment by yzffkindacrzy** [comment]
> It wouldn't have to disable all of the addons, just the ones with vulnerable code, for example.

Ahh, now I understand.

**Comment by WorldLanguages** [comment]
This would have been useful to fix #6062 more quickly

**Comment by Secret-chest** [comment]
If added, can it be an option? It feels pretty shady.

**Comment by mybearworld** [comment]
> If added, can it be an option? It feels pretty shady.

It's not going to be used for any shady purposes and is just for safety - we can (and should) be transparent about this, like adding a pop up saying something like "Example Addon was disabled because it had a major bug that deleted your account when pressing the space bar".

**Comment by Secret-chest** [comment]
You didn't answer my question.

**Comment by mybearworld** [comment]
I'm not in a position to answer it, but in my opinion, an option would do more harm than good.

**Comment by Secret-chest** [comment]
it's forkin' time

**Comment by mxmou** [comment]
I still don't like this idea. We have to be careful with any feature that involves sending requests in the background because some users will be concerned about privacy.

**Comment by GrahamSH-LLK** [comment]
Okay if this ever happens it's going to be super clear: a github repo set up with pages and a domain pointed to it. A single js file. Not behind cloudflare for extra transparency, users can check the cname themselves.

**Comment by WorldLanguages** [comment]
> I still don't like this idea. We have to be careful with any feature that involves sending requests in the background because some users will be concerned about privacy.

You may have a point, but I just want to bring up some points again, in case they were forgotten: 

- A server will be pinged only if Scratch is being used, so it's not really requesting "in the background". We may even send the requests from Scratch tabs if that results in cleaner code.
- The Privacy Policy already considers that scratchaddons.com might be reached automatically (and it already is, through `chrome.tabs.create()`, but it's visible to the user).
- There will be no middleman between the user and GitHub pages, like Graham just said.
- We could add an option to turn it off.

**Comment by Samq64** [comment]
I agree with having an opt-out toggle in more settings with a notice explaining why keeping it enabled is useful.

**Comment by WorldLanguages** [comment]
This becomes more important as the Edge Addons Store takes multiple days to review new version submissions.

**Comment by WorldLanguages** [comment]
> A server will be pinged only if Scratch is being used

PR #7505 intends to add this logic to the extension, but for other reasons. It can be reused to only ping our data server if Scratch (our our extension popup / settings page) is being actively used.

**Comment by WorldLanguages** [comment]
Here is how the setting could look like.

![image](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/a74209ce-ba5a-4914-a3ac-d3f6aca5a8c4)

**Comment by WorldLanguages** [comment]
Issue #7648 proved that even 0.7 requests per second per online user can effectively get close to a naive DDoS attack. So, as well as adding the ability to remotely disable addons, we should also add a general kill-switch to remotely disable the whole extension based on the version number.

**Comment by DNin01** [comment]
> So, as well as adding the ability to remotely disable addons, we should also add a general kill-switch to remotely disable the whole extension based on the version number.

I don't think we'd want or need to make an open-source extension as inert as a Spotify Car Thing. That's why we have patches.

**Comment by Hans5958** [comment]
Potential inspiration: https://github.com/refined-github/yolo

**Comment by DNin01** [comment]
Regarding privacy concerns...

I think it's definitely important, as an open-source extension, to collect data only when essential, and we haven't done anything like this before. So, if and when we add this, then for current users, I think we should pop up a screen informing them that we have added this function, what it means, and a link to our privacy policy, as well as giving them a choice to have it turned on or off. It would be off by default for existing users and potentially on by default for new users or we could give them a choice during setup.

The UI to ask for permission could say something like this:

> ## Scratch Addons can now periodically obtain data for updates and issues from scratchaddons.com.
> 
> This lets us announce issues and important messages to our user base, including you, if there are bugs, problems, or security vulnerabilities that we can't address right away. These messages will be displayed in the Scratch Addons popup and settings page.
>
> If you choose, we can make periodic requests to scratchaddons.com to get updates. This won't share any of your data with us.
> 
> <kbd>**That's great!**</kbd> <kbd>I'd rather not</kbd>
> 
> <sub>You can change this anytime in Scratch Addons' settings.</sub>
> 
> <sub>[Privacy policy](https://scratchaddons.com/docs/policies/privacy/extension/)</sub>

---

Related:
- #7531

**Comment by GrahamSH-LLK** [comment]
> Regarding privacy concerns...
> 
> I think it's definitely important, as an open-source extension, to collect data only when essential, and we haven't done anything like this before. So, if and when we add this, then for current users, I think we should pop up a screen informing them that we have added this function, what it means, and a link to our privacy policy, as well as giving them a choice to have it turned on or off. It would be off by default for existing users and potentially on by default for new users or we could give them a choice during setup.
> 
> The UI to ask for permission could say something like this:
> 
> > ## Scratch Addons can now periodically obtain data for updates and issues from scratchaddons.com.
> > This lets us announce issues and important messages to our user base, including you, if there are bugs, problems, or security vulnerabilities that we can't address right away. These messages will be displayed in the Scratch Addons popup and settings page.
> > If you choose, we can make periodic requests to scratchaddons.com to get updates. This won't share any of your data with us.
> > **That's great!** I'd rather not
> > You can change this anytime in Scratch Addons' settings.
> > [Privacy policy](https://scratchaddons.com/docs/policies/privacy/extension/)
> 
> Related:
> 
> * [Improve the flow of host permission granting #7531](https://github.com/ScratchAddons/ScratchAddons/issues/7531)

I.. don't think this is a privacy issue or really needs this. We _aren't_ collecting data, and this system is only useful if it's on for the vast majority of users. This has always been allowed under our [privacy policy](https://scratchaddons.com/docs/policies/privacy/extension/#third-parties).

**Comment by Secret-chest** [comment]
We will now have periodic pings. Not everyone wants to use a VPN just to not be fingerprinted by SA.

**Comment by mxmou** [comment]
> We will now have periodic pings. Not everyone wants to use a VPN just to not be fingerprinted by SA.

Fingerprinted in what way? If the extension fetches a JSON file from GitHub pages, no information about this request is stored anywhere. I think the only problem with the suggestion is that users might *think* they're being tracked.

**Comment by Secret-chest** [comment]
It's still an antifeature. Antifeatures are typical of proprietary software. When you need Pi-hole or modifications to the extension to use your addons, it's a problem.

**Comment by CST1229** [comment]
What if we just include the blog in the popup, as an addon (and then an option to show in the badge if there are unread posts)?

**Comment by Samq64** [comment]
> We will now have periodic pings. Not everyone wants to use a VPN just to not be fingerprinted by SA.

I don't like data collection either but this would be a GET request for a JSON file, not a POST. Also It would be opt-out and we already do things some users consider tracking such as opening ScratchAddons.com with a UTM on install and uninstall and adding an S to the flash version on the forums. There are also addons which contact ScratchDB, Ocular and TurboWarp etc.

**Comment by GarboMuffin** [comment]
The fingerprinting threat model is that someone who can monitor network traffic will see the plaintext SNI in the requests to `scratchaddons.github.io` or `whatever.scratchaddons.com` and conclude that you are probably a Scratch Addons user. Whether that is worth worrying about is, of course, debtable (if you can't trust your network you ought to be using a VPN already...).

**Comment by DNin01** [comment]
This is once again relevant with issues like these:
- #8708
- #8677

**Comment by Samq64** [comment]
Another option could be a local crash handler that offers to disable a broken addon. Of course it wouldn't cover security issues and has its own potenial problems such as false positives or disaplying many crash popups at once.

**Comment by WorldLanguages** [comment]
> Another option could be a local crash handler that offers to disable a broken addon.

Sounds infeasible.

**Comment by DNin01** [comment]
> Another option could be a local crash handler that offers to disable a broken addon. Of course it wouldn't cover security issues and has its own potenial problems such as false positives or disaplying many crash popups at once.

I could only imagine it…

**Comment by mxmou** [comment]
> > Another option could be a local crash handler that offers to disable a broken addon.
> 
> Sounds infeasible.

Scratch crashes are usually caused by an exception thrown inside an overridden Scratch function. If every addon that patches Scratch code used an addon API to do so, that API could wrap the overridden code in a try-catch block so that it could tell the user which addon failed.

The discuss-button crash happened because the addon was deleting elements rendered by React, which is usually a bad idea.

**Comment by DNin01** [comment]
> Scratch crashes are usually caused by an exception thrown inside an overridden Scratch function. If every addon that patches Scratch code used an addon API to do so, that API could wrap the overridden code in a try-catch block so that it could tell the user which addon failed.

I was thinking that too! But what happens if an exception is caught? Do we fall back and call the original function? Do we stop execution? Do we change any settings?

**Comment by mxmou** [comment]
> I was thinking that too! But what happens if an exception is caught? Do we fall back and call the original function? Do we stop execution? Do we change any settings?

We could try calling the original function, but that might not help if the addon's code left things in a broken state. I don't think settings should be changed automatically - we can simply tell the user which addon failed so they can disable it.

**Comment by DNin01** [comment]
> we can simply tell the user which addon failed so they can disable it.

For this, we could pop an alert over the page, but that could be annoying if the error is unrelated, or we could show the error in the extension's popup, but it might not be seen by the user for a while.
